doozer/doozerlib/distgit.py

```diff
@@ -1528,66 +1584,26 @@ def _should_match_upstream(self) -> bool:
-            cmd = f'oc image info {original_parent} -o json'
-            out, _ = exectools.cmd_assert(cmd, retries=3)
-            labels = json.loads(out)['config']['config']['Labels']
-
-            # Get the exact build NVR
-            build_nvr = f'{labels["com.redhat.component"]}-{labels["version"]}-{labels["release"]}'
-
-            # Query Brew for build info
-            self.logger.debug('Retrieving info for Brew build %s', build_nvr)
-            with self.runtime.shared_koji_client_session() as koji_api:
-                if not koji_api.logged_in:
-                    koji_api.gssapi_login()
-                build = koji_api.getBuild(build_nvr, strict=True)
-
-            # Get the pullspec for the upstream equivalent
-            upstream_equivalent_pullspec = build['extra']['image']['index']['pull'][1]
-
-            # Verify whether the image exists
-            self.logger.debug('Checking for upstream equivalent existence, pullspec: %s', upstream_equivalent_pullspec)
-            cmd = f'oc image info {upstream_equivalent_pullspec} --filter-by-os linux/amd64 -o json'
-            out, _ = exectools.cmd_assert(cmd, retries=3)
-
-            # It does. Use this to rebase FROM directive
-            digest = json.loads(out)['digest']
-
-            mapped_image = f'{labels["name"]}@{digest}'
-            # if upstream equivalent does not match ART's config, add a warning to the Dockerfile
-            if mapped_image != stream.image:
-                dfp.add_lines_at(
-                    0,
-                    "",
-                    "# Parent images were rebased matching upstream equivalent that didn't match ART's config",
-                    "")
-                self.logger.info('Will override %s with upsteam equivalent %s', stream.image, mapped_image)
-            return mapped_image
-
-        except (KeyError, ChildProcessError) as e:
-            # We could get:
-            #   - a ChildProcessError when the upstream equivalent is not found
-            #   - a ChildProcessError when trying to rebase our base images
-            #   - a KeyError when 'com.redhat.component' label is undefined
-            # In all of the above, we'll just do typical stream resolution
-
-            self.logger.warning(f'Could not match upstream parent {original_parent}: {e}')
-            dfp.add_lines_at(
-                0,
-                "",
-                "# Failed matching upstream equivalent, ART configuration was used to rebase parent images",
-                ""
-            )
+        parent = self._resolve_parent(original_parent, dfp)
+        if not parent:
             return stream.image
+        else:
+            return parent
 
     def _rebase_from_directives(self, dfp):
+        self.should_match_upstream = self._should_match_upstream()
+        if self.should_match_upstream:
+            # ART-8476 override image config with upstream
+            alt_upstream = self.config.get('alternative_upstream', None)
```

Comment:
If we're using https://gist.github.com/jupierce/1e9eaae40a47a3fd097d5d4fbcf3b93c#file-gistfile1-txt-L40-L44 as an example, it has a `- when: el8`, we would want to validate that here, since upstream could be using rhel9 but this would mean we use rhel8 config.

Reply:
We can also separate out implementing this alt_upstream change in another PR, and let this PR through

Reply:
I think for the moment we can just an alternative config we use when `canonical_builders` is True, and blindly apply it if needed. To get more flexible, we sure need a couple of `when` clauses, but we need to implement a way to determine upstream rhel version. I'd do this as a separate ticket